# Lab | Data Aggregation and Filtering

In this challenge, we will continue to work with customer data from an insurance company. We will use the dataset called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by first performing data cleaning, formatting, and structuring.

In [1]:
import pandas as pd
url = ("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv")
df = pd.read_csv(url)

In [2]:
display(df.head())

Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
0           0  DK49336     Arizona              4809.216960       No   
1           1  KX64629  California              2228.525238       No   
2           2  LZ68649  Washington             14947.917300       No   
3           3  XL78013      Oregon             22332.439460      Yes   
4           4  QA50777      Oregon              9025.067525       No   

   Coverage Education Effective To Date EmploymentStatus Gender  ...  \
0     Basic   College           2/18/11         Employed      M  ...   
1     Basic   College           1/18/11       Unemployed      F  ...   
2     Basic  Bachelor           2/10/11         Employed      M  ...   
3  Extended   College           1/11/11         Employed      M  ...   
4   Premium  Bachelor           1/17/11    Medical Leave      F  ...   

   Number of Open Complaints Number of Policies     Policy Type        Policy  \
0                        0.0                  9  Corporate Auto  Corporate L3   
1                        0.0                  1   Personal Auto   Personal L3   
2                        0.0                  2   Personal Auto   Personal L3   
3                        0.0                  2  Corporate Auto  Corporate L3   
4                        NaN                  7   Personal Auto   Personal L2   

   Renew Offer Type  Sales Channel  Total Claim Amount  Vehicle Class  \
0            Offer3          Agent          292.800000  Four-Door Car   
1            Offer4    Call Center          744.924331  Four-Door Car   
2            Offer3    Call Center          480.000000            SUV   
3            Offer2         Branch          484.013411  Four-Door Car   
4            Offer1         Branch          707.925645  Four-Door Car   

  Vehicle Size Vehicle Type  
0      Medsize          NaN  
1      Medsize          NaN  
2      Medsize            A  
3      Medsize            A  
4      Medsize          NaN  

[5 rows x 26 columns]

In [3]:
df.columns

Index(['Unnamed: 0', 'Customer', 'State', 'Customer Lifetime Value',
       'Response', 'Coverage', 'Education', 'Effective To Date',
       'EmploymentStatus', 'Gender', 'Income', 'Location Code',
       'Marital Status', 'Monthly Premium Auto', 'Months Since Last Claim',
       'Months Since Policy Inception', 'Number of Open Complaints',
       'Number of Policies', 'Policy Type', 'Policy', 'Renew Offer Type',
       'Sales Channel', 'Total Claim Amount', 'Vehicle Class', 'Vehicle Size',
       'Vehicle Type'],
      dtype='object')

In [5]:
df.columns = ['Unnamed: 0', 'Customer', 'State', 'Customer Lifetime Value',
       'Response', 'Coverage', 'Education', 'Effective To Date',
       'EmploymentStatus', 'Gender', 'Income', 'Location Code',
       'Marital Status', 'Monthly Premium Auto', 'Months Since Last Claim',
       'Months Since Policy Inception', 'Number of Open Complaints',
       'Number of Policies', 'Policy Type', 'Policy', 'Renew Offer Type',
       'Sales Channel', 'Total Claim Amount', 'Vehicle Class', 'Vehicle Size',
       'Vehicle Type']

In [6]:
df.rename(columns={col: col.replace(" ", "_").lower() for col in df.columns}, inplace=True)

In [7]:
display(df)

unnamed:_0 customer       state  customer_lifetime_value response  \
0               0  DK49336     Arizona              4809.216960       No   
1               1  KX64629  California              2228.525238       No   
2               2  LZ68649  Washington             14947.917300       No   
3               3  XL78013      Oregon             22332.439460      Yes   
4               4  QA50777      Oregon              9025.067525       No   
...           ...      ...         ...                      ...      ...   
10905       10905  FE99816      Nevada             15563.369440       No   
10906       10906  KX53892      Oregon              5259.444853       No   
10907       10907  TL39050     Arizona             23893.304100       No   
10908       10908  WA60547  California             11971.977650       No   
10909       10909  IV32877         NaN              6857.519928      NaN   

       coverage education effective_to_date employmentstatus gender  ...  \
0         Basic   College           2/18/11         Employed      M  ...   
1         Basic   College           1/18/11       Unemployed      F  ...   
2         Basic  Bachelor           2/10/11         Employed      M  ...   
3      Extended   College           1/11/11         Employed      M  ...   
4       Premium  Bachelor           1/17/11    Medical Leave      F  ...   
...         ...       ...               ...              ...    ...  ...   
10905   Premium  Bachelor           1/19/11       Unemployed      F  ...   
10906     Basic   College            1/6/11         Employed      F  ...   
10907  Extended  Bachelor            2/6/11         Employed      F  ...   
10908   Premium   College           2/13/11         Employed      F  ...   
10909     Basic  Bachelor            1/8/11       Unemployed      M  ...   

       number_of_open_complaints number_of_policies     policy_type  \
0                            0.0                  9  Corporate Auto   
1                            0.0                  1   Personal Auto   
2                            0.0                  2   Personal Auto   
3                            0.0                  2  Corporate Auto   
4                            NaN                  7   Personal Auto   
...                          ...                ...             ...   
10905                        NaN                  7   Personal Auto   
10906                        0.0                  6   Personal Auto   
10907                        0.0                  2  Corporate Auto   
10908                        4.0                  6   Personal Auto   
10909                        0.0                  3   Personal Auto   

             policy  renew_offer_type  sales_channel  total_claim_amount  \
0      Corporate L3            Offer3          Agent          292.800000   
1       Personal L3            Offer4    Call Center          744.924331   
2       Personal L3            Offer3    Call Center          480.000000   
3      Corporate L3            Offer2         Branch          484.013411   
4       Personal L2            Offer1         Branch          707.925645   
...             ...               ...            ...                 ...   
10905   Personal L1            Offer3            Web         1214.400000   
10906   Personal L3            Offer2         Branch          273.018929   
10907  Corporate L3            Offer1            Web          381.306996   
10908   Personal L1            Offer1         Branch          618.288849   
10909   Personal L1            Offer4            Web         1021.719397   

       vehicle_class vehicle_size vehicle_type  
0      Four-Door Car      Medsize          NaN  
1      Four-Door Car      Medsize          NaN  
2                SUV      Medsize            A  
3      Four-Door Car      Medsize            A  
4      Four-Door Car      Medsize          NaN  
...              ...          ...          ...  
10905     Luxury Car      Medsize            A  
10906  Four-Door Car   

In [10]:
df["customer_lifetime_value"] = df["customer_lifetime_value"].round(2)

In [17]:
display(df)

unnamed:_0 customer       state  customer_lifetime_value response  \
0               0  DK49336     Arizona                  4809.22       No   
1               1  KX64629  California                  2228.53       No   
2               2  LZ68649  Washington                 14947.92       No   
3               3  XL78013      Oregon                 22332.44      Yes   
4               4  QA50777      Oregon                  9025.07       No   
...           ...      ...         ...                      ...      ...   
10905       10905  FE99816      Nevada                 15563.37       No   
10906       10906  KX53892      Oregon                  5259.44       No   
10907       10907  TL39050     Arizona                 23893.30       No   
10908       10908  WA60547  California                 11971.98       No   
10909       10909  IV32877         NaN                  6857.52      NaN   

       coverage education effective_to_date employmentstatus gender  ...  \
0         Basic   College           2/18/11         Employed      M  ...   
1         Basic   College           1/18/11       Unemployed      F  ...   
2         Basic  Bachelor           2/10/11         Employed      M  ...   
3      Extended   College           1/11/11         Employed      M  ...   
4       Premium  Bachelor           1/17/11    Medical Leave      F  ...   
...         ...       ...               ...              ...    ...  ...   
10905   Premium  Bachelor           1/19/11       Unemployed      F  ...   
10906     Basic   College            1/6/11         Employed      F  ...   
10907  Extended  Bachelor            2/6/11         Employed      F  ...   
10908   Premium   College           2/13/11         Employed      F  ...   
10909     Basic  Bachelor            1/8/11       Unemployed      M  ...   

       number_of_open_complaints number_of_policies     policy_type  \
0                            0.0                  9  Corporate Auto   
1                            0.0                  1   Personal Auto   
2                            0.0                  2   Personal Auto   
3                            0.0                  2  Corporate Auto   
4                            NaN                  7   Personal Auto   
...                          ...                ...             ...   
10905                        NaN                  7   Personal Auto   
10906                        0.0                  6   Personal Auto   
10907                        0.0                  2  Corporate Auto   
10908                        4.0                  6   Personal Auto   
10909                        0.0                  3   Personal Auto   

             policy  renew_offer_type  sales_channel  total_claim_amount  \
0      Corporate L3            Offer3          Agent              292.80   
1       Personal L3            Offer4    Call Center              744.92   
2       Personal L3            Offer3    Call Center              480.00   
3      Corporate L3            Offer2         Branch              484.01   
4       Personal L2            Offer1         Branch              707.93   
...             ...               ...            ...                 ...   
10905   Personal L1            Offer3            Web             1214.40   
10906   Personal L3            Offer2         Branch              273.02   
10907  Corporate L3            Offer1            Web              381.31   
10908   Personal L1            Offer1         Branch              618.29   
10909   Personal L1            Offer4            Web             1021.72   

       vehicle_class vehicle_size vehicle_type  
0      Four-Door Car      Medsize          NaN  
1      Four-Door Car      Medsize          NaN  
2                SUV      Medsize            A  
3      Four-Door Car      Medsize            A  
4      Four-Door Car      Medsize          NaN  
...              ...          ...          ...  
10905     Luxury Car      Medsize            A  
10906  Four-Door Car   

1. Create a new DataFrame that only includes customers who have a total_claim_amount greater than $1,000 and have a response of "Yes" to the last marketing campaign.

In [19]:
filtered_df = df[(df['total_claim_amount'] > 1000) & (df['response'] == 'Yes')]
filtered_df

unnamed:_0 customer       state  customer_lifetime_value response  \
189           189  OK31456  California                 11009.13      Yes   
236           236  YJ16163      Oregon                 11009.13      Yes   
419           419  GW43195      Oregon                 25807.06      Yes   
442           442  IP94270     Arizona                 13736.13      Yes   
587           587  FJ28407  California                  5619.69      Yes   
...           ...      ...         ...                      ...      ...   
10351       10351  FN44127      Oregon                  3508.57      Yes   
10373       10373  XZ64172      Oregon                 10963.96      Yes   
10487       10487  IX60941      Oregon                  3508.57      Yes   
10565       10565  QO62792      Oregon                  7840.17      Yes   
10708       10708  CK39096      Oregon                  5619.69      Yes   

       coverage             education effective_to_date employmentstatus  \
189     Premium              Bachelor           1/24/11         Employed   
236     Premium              Bachelor           1/24/11         Employed   
419    Extended               College           2/13/11         Employed   
442     Premium                Master           2/13/11         Disabled   
587     Premium  High School or Below           1/26/11       Unemployed   
...         ...                   ...               ...              ...   
10351  Extended               College            1/5/11    Medical Leave   
10373   Premium  High School or Below            2/8/11         Employed   
10487  Extended               College            1/5/11    Medical Leave   
10565  Extended               College           1/14/11         Employed   
10708   Premium  High School or Below           1/26/11       Unemployed   

      gender  ...  number_of_open_complaints number_of_policies  \
189        F  ...                        0.0                  1   
236        F  ...                        0.0                  1   
419        F  ...                        1.0                  2   
442        F  ...                        0.0                  8   
587        M  ...                        0.0                  1   
...      ...  ...                        ...                ...   
10351      M  ...                        1.0                  1   
10373      M  ...                        0.0                  1   
10487      M  ...                        1.0                  1   
10565      M  ...                        2.0                  1   
10708      M  ...                        0.0                  1   

          policy_type        policy  renew_offer_type  sales_channel  \
189    Corporate Auto  Corporate L3            Offer2          Agent   
236      Special Auto    Special L3            Offer2          Agent   
419     Personal Auto   Personal L2            Offer1         Branch   
442     Personal Auto   Personal L2            Offer1            Web   
587     Personal Auto   Personal L1            Offer2            Web   
...               ...           ...               ...            ...   
10351   Personal Auto   Personal L2            Offer2         Branch   
10373  Corporate Auto  Corporate L2            Offer1          Agent   
10487   Personal Auto   Personal L3            Offer2         Branch   
10565   Personal Auto   Personal L3            Offer2          Agent   
10708   Personal Auto   Personal L3            Offer2            Web   

       total_claim_amount  vehicle_class vehicle_size vehicle_type  
189               1358.40     Luxury Car      Medsize          NaN  
236               1358.40     Luxury Car      Medsize            A  
419               1027.20     Luxury Car        Small            A  
442               1261.32            SUV      Medsize            A  
587               1027.00            SUV      Medsize            A  
...                   ...            ...          ...          ...  
10351             1176.28  F

2. Using the original Dataframe, analyze the average total_claim_amount by each policy type and gender for customers who have responded "Yes" to the last marketing campaign. Write your conclusions.

In [20]:
df_filtered2 = df[df['response'] == 'Yes']

In [23]:
display (df_filtered2)

unnamed:_0 customer       state  customer_lifetime_value response  \
3               3  XL78013      Oregon                 22332.44      Yes   
8               8  FM55990  California                  5989.77      Yes   
15             15  CW49887  California                  4626.80      Yes   
19             19  NJ54277  California                  3746.75      Yes   
27             27  MQ68407      Oregon                  4376.36      Yes   
...           ...      ...         ...                      ...      ...   
10844       10844  FM31768     Arizona                  5979.72      Yes   
10852       10852  KZ80424  Washington                  8382.48      Yes   
10872       10872  XT67997  California                  5979.72      Yes   
10887       10887  BY78730      Oregon                  8879.79      Yes   
10897       10897  MM70762     Arizona                  9075.77      Yes   

       coverage             education effective_to_date employmentstatus  \
3      Extended               College           1/11/11         Employed   
8       Premium               College           1/19/11         Employed   
15        Basic                Master           1/16/11         Employed   
19     Extended               College           2/26/11         Employed   
27      Premium              Bachelor           2/28/11         Employed   
...         ...                   ...               ...              ...   
10844  Extended  High School or Below            2/7/11         Employed   
10852     Basic              Bachelor           1/27/11         Employed   
10872  Extended  High School or Below            2/7/11         Employed   
10887     Basic  High School or Below            2/3/11         Employed   
10897     Basic                Master           1/26/11         Employed   

      gender  ...  number_of_open_complaints number_of_policies  \
3          M  ...                        0.0                  2   
8          M  ...                        0.0                  1   
15         F  ...                        0.0                  1   
19         F  ...                        1.0                  1   
27         F  ...                        0.0                  1   
...      ...  ...                        ...                ...   
10844      F  ...                        0.0                  3   
10852      M  ...                        0.0                  2   
10872      F  ...                        0.0                  3   
10887      F  ...                        0.0                  7   
10897      M  ...                        0.0                  8   

          policy_type        policy  renew_offer_type  sales_channel  \
3      Corporate Auto  Corporate L3            Offer2         Branch   
8       Personal Auto   Personal L1            Offer2         Branch   
15       Special Auto    Special L1            Offer2         Branch   
19      Personal Auto   Personal L2            Offer2    Call Center   
27      Personal Auto   Personal L3            Offer2          Agent   
...               ...           ...               ...            ...   
10844   Personal Auto   Personal L1            Offer2          Agent   
10852   Personal Auto   Personal L2            Offer2    Call Center   
10872   Personal Auto   Personal L3            Offer2          Agent   
10887    Special Auto    Special L2            Offer1          Agent   
10897   Personal Auto   Personal L1            Offer1          Agent   

       total_claim_amount  vehicle_class vehicle_size vehicle_type  
3                  484.01  Four-Door Car      Medsize            A  
8                  739.20     Sports Car      Medsize          NaN  
15                 547.20            SUV      Medsize          NaN  
19                  19.58   Two-Door Car        Large            A  
27                  60.04  Four-Door Car      Medsize          NaN  
...                   ...            ...          ...          ...  
10844              547.20  F

In [26]:
df_filtered2.info ()

<class 'pandas.core.frame.DataFrame'>
Index: 1466 entries, 3 to 10897
Data columns (total 26 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   unnamed:_0                     1466 non-null   int64  
 1   customer                       1466 non-null   object 
 2   state                          1466 non-null   object 
 3   customer_lifetime_value        1466 non-null   float64
 4   response                       1466 non-null   object 
 5   coverage                       1466 non-null   object 
 6   education                      1466 non-null   object 
 7   effective_to_date              1466 non-null   object 
 8   employmentstatus               1466 non-null   object 
 9   gender                         1466 non-null   object 
 10  income                         1466 non-null   int64  
 11  location_code                  1466 non-null   object 
 12  marital_status                 1466 non-null   objec

In [30]:
import numpy as np

In [37]:
df_filtered2.groupby(['gender', 'policy_type'])['total_claim_amount'].mean()

gender  policy_type   
F       Corporate Auto    433.738462
        Personal Auto     452.966037
        Special Auto      453.279714
M       Corporate Auto    408.582468
        Personal Auto     457.010149
        Special Auto      429.528125
Name: total_claim_amount, dtype: float64

# Gender Differences:

Female Customers:
On average, female customers have higher total claim amounts compared to male customers for corporate and special auto policies. The average total claim amounts for female customers are 433.74 for corporate auto, 452.97 for personal auto, and 453.28 for special auto.

Male Customers:
Male customers have a slightly higher average total claim amount for personal auto policies at 457.01, but lower for corporate auto (408.58) and special auto (429.53) compared to females.

# Policy Type Impact:

Female customers:
The claim amounts are relatively similar across different policy types, with a slightly higher average for special auto policies.
Male customers:
Personal auto policies have the highest average total claim amount, indicating a trend where claims under personal auto policies are higher for men.

# Overall Insights:

The claim amounts are fairly consistent across genders and policy types, with personal auto policies yielding slightly higher average claim amounts, particularly for male customers.
There's a noticeable gender difference in claim amounts for corporate auto policies, with female customers filing higher claims on average than male customers. This could be indicative of different usage patterns, risk profiles, or vehicle types insured under these policies.

3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers.

In [63]:
state_customer_count = df['state'].value_counts()
states_more_than_500_customers = state_customer_count[state_customer_count > 500]
display(states_more_than_500_customers)

state
California    3552
Oregon        2909
Arizona       1937
Nevada         993
Washington     888
Name: count, dtype: int64

# I am not sure that this is the supposed output...

In summary, California, Oregon, Arizona, Nevada, and Washington are key markets for the insurance company, with each state hosting a significant number of policyholders. Strategies for these states should focus on strengthening market presence, enhancing customer relationships, and exploring avenues for further growth and efficiency improvements

4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.

In [69]:
agg_results = df.groupby(['gender', 'education']).agg({
    'customer_lifetime_value': ['max', 'min', 'median']})
display(agg_results)

customer_lifetime_value                   
                                                max      min    median
gender education                                                      
F      Bachelor                            73225.96  1904.00  5640.505
       College                             61850.19  1898.68  5623.610
       Doctor                              44856.11  2395.57  5332.460
       High School or Below                55277.45  2144.92  6039.555
       Master                              51016.07  2417.78  5729.860
M      Bachelor                            67907.27  1898.01  5548.030
       College                             61134.68  1918.12  6005.850
       Doctor                              32677.34  2267.60  5577.670
       High School or Below                83325.38  1940.98  6286.735
       Master                              50568.26  2272.31  5579.100

The provided data outlines the maximum, minimum, and median customer lifetime value, segmented by gender and education level. Analyzing this data reveals several insights into the demographics of an insurer's customer base and their potential value to the company.

# Gender and Education

Highest customer lifetime value:
The highest maximum is observed in male customers with a High School or Below education level at 83,325.38.

Lowest customer lifetime value:
The lowest minimum is found in female customers with a Bachelor's degree at 1,904.00.

# Median

Female Customers:
The median customer lifetime value is relatively consistent across different education levels, with High School or Below having the highest median with 6,039.555.
Bachelor and College-educated female customers have slightly lower median customer lifetime values.

Male Customers:
High School or Below education have the highest median customer lifetime value with 6,286.735, closely followed by those with a College education. 
The Doctor category shows the lowest maximum customer lifetime value for males.

# Education Level Variation:

Doctorate Level: Notably, individuals with a Doctorate show a significant difference in maximum customer lifetime value between genders, with females peaking at 44,856.11 and males at a much lower 32,677.34.

Master’s Degree: Both genders show similar patterns in median customer lifetime value for Master’s degree holders.

## Bonus

5. The marketing team wants to analyze the number of policies sold by state and month. Present the data in a table where the months are arranged as columns and the states are arranged as rows.

6.  Display a new DataFrame that contains the number of policies sold by month, by state, for the top 3 states with the highest number of policies sold.

*Hint:*
- *To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group. Afterwards, you will need to sort the data by the count of policies sold in descending order.*
- *Next, you will select the top 3 states with the highest number of policies sold.*
- *Finally, you will create a new DataFrame that contains the number of policies sold by month for each of the top 3 states.*

7. The marketing team wants to analyze the effect of different marketing channels on the customer response rate.

Hint: You can use melt to unpivot the data and create a table that shows the customer response rate (those who responded "Yes") by marketing channel.

External Resources for Data Filtering: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

In [ ]:
# your code goes here